oven-sh/bun
# fix syscall error handling

Hello,

first of all I want to express that I'm new to bun. I've tried to built an SvelteKit application and ran into the same issue as #4723. Afterwards I jumped down deep into that rabbit hole :-D 

Fixes #4723

Possible related to some extend to one or more of #4671 #5935 #5943 #4896 #5306 #1446 

### What does this PR do?

The error handling between calling zig syscalls and using libc is different. Within zigs implementation, and therefore it's clone within bun, the return value of a syscall is of type usize. Libc implemenations use a signed integer namely c_int. As soon as libc is linked every response code of a syscall is interpreted via the libc errno function, but not every syscall is made through the libc interface. Only a few syscalls on aarch64 are used from libc every other one is called through zig itself. This could lead to errors beeing misinterpreted as success, cause its strictly positive and therefore not an error for the libc implementation. If an open syscall internally fails but is interpreted as a success the following write syscall would operate on a wrong file descriptor resulting in error bad file descriptor (EBADF).

### How did you verify your code works?

I've built a release version of bun and verified that the expected behaviour is observed.
Before that I've built a reproducer to force the bad behaviour.

1. Create a _index.ts_ with the below content.
2. Create a directory called test `mkdir test`
3. Run bun via `bun index.ts`
4. This should result in ISDIR error, cause the file gets opened an the syscall fails with _ERRISDIR_ but instead the open succeeds and the stat afterwards gets an _EBADF_.

# TypeScript reproducer

```TypeScript
import fs from "fs";
fs.writeFileSync(`test`, `hello world`);
```

The error from the second syscall is correctly interpreted as this is run on aarch64 and bun uses libc stat on aarch64.

https://github.com/oven-sh/bun/blob/6301778a589254e2c3c0d95f768fce303f528b03/src/sys.zig#L31-L50

## Before
```
# ./packages/bun-linux-aarch64/bun index.ts
1 | import fs from "fs";
2 | fs.writeFileSync(`test`, `hello world`);
   ^
EBADF: Bad file descriptor
    fd: "-21"
 syscall: "write"
   errno: -9

      at /home/user/tmp/bun/index.ts:2:0
```

## After
```
# ./packages/bun-linux-aarch64/bun index.ts
1 | import fs from "fs";
2 | fs.writeFileSync(`test`, `hello world`);
   ^
EISDIR: Is a directory
   path: "test"
 syscall: "open"
   errno: -21

      at /home/user/tmp/bun/index.ts:2:0
```

# Zig reproducer

```Zig
const std = @import("std");
const os = std.os;
const linux = os.linux;
const system = linux;
const builtin = @import("builtin");

pub fn main() !void {
    std.log.info("libc {}", .{builtin.link_libc});
    const rc_libc = system.openat(@as(system.fd_t, @intCast(-100)), "test", 321, 438);
    switch (os.errno(rc_libc)) {
        .SUCCESS => std.log.info("success", .{}),
        .ISDIR => std.log.info("is directory", .{}),
        else => return,
    }

    const rc_linux = system.openat(@as(system.fd_t, @intCast(-100)), "test", 321, 438);
    switch (linux.getErrno(rc_linux)) {
        .SUCCESS => std.log.info("success", .{}),
        .ISDIR => std.log.info("is directory", .{}),
        else => return,
    }
}
```

## Output after building with _zig build-exe main.zig_
```
info: libc false
info: is directory
info: is directory
```

## Output after building with _zig build-exe main.zig --library c_
```
info: libc true
info: success
info: is directory
```
